code/game/machinery/engi_points_vendor.dm

```diff
@@ -0,0 +1,674 @@
+		new /datum/track("Engineering's Superiority Shimmy", 			'sound/misc/e1m1.ogg', 	1810, 	5),
+		new /datum/track("Engineering's Ultimate High-Energy Hustle",	'sound/misc/e1m1.ogg',	2260, 	7),
+		)
+	var/datum/track/selection = null
+
+/datum/track
+	var/song_name = "generic"
+	var/song_path = null
+	var/song_length = 0
+	var/song_beat = 0
+
+/datum/track/New(name, path, length, beat)
+	song_name = name
+	song_path = path
+	song_length = length
+	song_beat = beat
+
+/obj/machinery/disco/Initialize()
+	selection = songs[1]
+
+
+/obj/machinery/disco/Destroy()
+	dance_over()
+	return ..()
+
+/obj/machinery/disco/attackby(obj/item/O, mob/user, params)
+	if(!active)
+		if(istype(O, /obj/item/weapon/wrench))
+			if(!anchored && !isinspace())
+				user << "<span class='notice'>You secure the [src] to the floor.</span>"
+				anchored = TRUE
+			else if(anchored)
+				user << "<span class='notice'>You unsecure and disconnect the [src].</span>"
+				anchored = FALSE
+			playsound(loc, 'sound/items/Deconstruct.ogg', 50, 1)
+			return
+	return ..()
+
+
+/obj/machinery/disco/interact(mob/user)
+	if (!anchored)
+		user << "<span class='warning'>This device must be anchored by a wrench!</span>"
+		return
+	if(!allowed(user))
+		user << "<span class='warning'>You are overwhelmed by the raw amount of data being displayed, only an engineer could operate such a sophisticated device.</span>"
+		playsound(src, 'sound/misc/compiler-failure.ogg', 50, 1)
+		return
+	if(!Adjacent(user))
+		if(!isAI(user))
+			return
+	user.set_machine(src)
+	var/list/dat = list()
+	dat +="<div class='statusDisplay' style='text-align:center'>"
+	dat += "<b><A href='?src=\ref[src];action=toggle'>[!active ? "BREAK IT DOWN" : "SHUT IT DOWN"]<b></A><br>"
+	dat += "</div><br>"
+	dat += "<A href='?src=\ref[src];action=select'> Select Track</A><br>"
+	dat += "Track Selected: [selection.song_name]<br>"
+	dat += "Track Length: [selection.song_length/10] seconds<br><br>"
+	dat += "<i>More songs can be unlocked by earning more IEV points</i><br>"
+	dat += "<br>DJ's Soundboard:<b><br>"
+	dat +="<div class='statusDisplay'><div style='text-align:center'>"
+	dat += "<A href='?src=\ref[src];action=horn'>Air Horn</A>  "
+	dat += "<A href='?src=\ref[src];action=alert'>Station Alert</A>  "
+	dat += "<A href='?src=\ref[src];action=siren'>Warning Siren</A>  "
+	dat += "<A href='?src=\ref[src];action=honk'>Honk</A><br>"
+	dat += "<A href='?src=\ref[src];action=pump'>Shotgun Pump</A>"
+	dat += "<A href='?src=\ref[src];action=pop'>Gunshot</A>"
+	dat += "<A href='?src=\ref[src];action=saber'>Esword</A>"
+	dat += "<A href='?src=\ref[src];action=harm'>Harm Alarm</A>"
+	var/datum/browser/popup = new(user, "vending", "Radiance Dance Machine - Mark IV", 400, 350)
+	popup.set_content(dat.Join())
+	popup.open()
+
+
+/obj/machinery/disco/Topic(href, href_list)
+	if(..())
+		return
+	add_fingerprint(usr)
+	switch(href_list["action"])
+		if("toggle")
+			if (!src || QDELETED(src))
+				return
```

Comment:
I'm fairly certain this is checked in ..()

Also `!src` is checked in `QDELETED`